### ISSUE TITLE
DATETIME(2)FROMPARTS

### ISSUE THREAD

**Issue by sd-mobile** [linq2db/linq2db#3644]
DATETIME(2)FROMPARTS not working in MERGE statement

Sql error Msg 10760, Level 16, State 1, Line 17
Scale argument is not valid. Valid expressions for data type datetime2 scale argument are integer constants and integer constant expressions.

Generated SQL query:

```sql
MERGE INTO [FIN].[agreement] [Target]
USING (
VALUES
(
	'279f50cd-7e63-4169-ed20-12504ddbce50',
	DATETIME2FROMPARTS(2022, 7, 2, 0, 0, 0, 0, 0)
)
)
[Source]
(
	[dbid],         -- UNIQUEIDENTIFIER
	[signDate]   -- DATETIME2(0)
)

ON ([Target].[dbid] = [Source].[dbid])

WHEN MATCHED THEN
UPDATE
SET
	[Target].[signDate] = [Source].[signDate]
```

### Environment details
linq2db version: 4.1
Database Server: SqlServer 2017
Database Provider: SqlServer.2012 or above
Operating system: Windows 10
.NET Framework: 6.0

Maybe add the ability to disable the use of DATETIME(2)FROMPARTS?

**Comment by MaceWindu** [comment]
could you provide table schema? I cannot reproduce issue and don't see how it even possible to get such error with `DATETIME2FROMPARTS(2022, 7, 2, 0, 0, 0, 0, 0)`

**Comment by sd-mobile** [comment]
This code reproduce same error
SqlServerVersion = SqlServerVersion.v2012
SqlServerTools.Provider = SqlServerProvider.MicrosoftDataSqlClient

        [Table(Name = "test_agreement")]
        private class TEST_agreement
        {
            [PrimaryKey]
            public Guid dbid { get; set; }
            [Column(DataType = DataType.DateTime2, Precision = 0)]
            public DateTime? signDate { get; set; }
        }

                using var tmpTable = mainDb.CreateTempTable<TEST_agreement>();

                var items = Enumerable
                    .Range(0, 1)
                    .Select
                    (
                        f =>  new TEST_agreement
                        {
                            dbid = Guid.NewGuid(),
                            signDate = DateTime.Now,
                        }
                    )
                    .ToArray();
                tmpTable.Merge()
                    .Using(items)
                    .OnTargetKey()
                    .InsertWhenNotMatched()
                    .UpdateWhenMatched()
                    .Merge();

**Comment by MaceWindu** [comment]
Will check. Could you also provide exact sql server version (`select @@version`)?

**Comment by sd-mobile** [comment]
Microsoft SQL Server 2014 (SP2-GDR) (KB4505217) - 12.0.5223.6 (X64)   May 26 2019 20:36:50   Copyright (c) Microsoft Corporation  Standard Edition (64-bit) on Windows NT 6.3 <X64> (Build 9600: ) (Hypervisor)

**Comment by MaceWindu** [comment]
Still cannot reproduce.

Two options:
- there is something missing from your test
- Sql Server bug. Your version miss a lot of updates https://support.microsoft.com/en-us/topic/kb2936603-sql-server-2014-build-versions-6f75da99-d86f-53fa-23ce-3d2b4825eccb - there is SP3 + 4 CU available, but your version is SP2 without any CUs

I would recommend to update sql server. Or as workaround you can define own literal generation logic:
```cs
mappingSchema.SetValueToSqlConverter(typeof(DateTime),         (sb,dt,v) => CustomConvertDateTimeToSql(sb, dt.Type.DataType, v.ToString()!));
```
note that it overrides all existing `DateTime` mappings, so you will need to copy and edit existing [`ConvertDateTimeToSql`](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs) method and edit `datetime2` part

**Comment by sd-mobile** [comment]
Да, я тоже склоняюсь, что это ошибка в SqlServer
Конвертер применю, но тем не менее было бы не плохо иметь возможность отключать использование DATETIME(2)FROMPARTS более простым способом.
Не думаю, что я один напоролся на это.

**Comment by MaceWindu** [comment]
Requires #1181

**Comment by MaceWindu** [comment]
We don't plan any special "fix" for it as it is 99% non-updated database bug, which we don't support. Valid way to get this errror is to pass null as last parameter `DATETIME2FROMPARTS(2022, 7, 2, 0, 0, 0, 0, null)` ([see](http://www.sql-server-helper.com/error-messages/msg-10760.aspx)) but it is not possible in our code.

More generic fix already tracked by #1181 (hopefully part of v7 future release).
